oslo/torch/nn/parallel/data_parallel/data_parallel.py

```diff
@@ -0,0 +1,232 @@
+    )
+
+    add_wrapper(
+        module,
+        mode=ParallelMode.DATA,
+        wrapper=ddp,
+        parallel_context=parallel_context,
+    )
+    return module
+
+
+class _DistirbutedDataParallelWrapper(OsloParallelWrapper):
+    """Distributed data parallel wrapper for Oslo.
+    Example:
+        >>> from oslo.torch.nn.parallel import DistributedDataParallel as DDP
+        >>> model = torch.nn.Linear(20, 1)
+        >>>
+        >>> model = DDP(model, parallel_context)
+        >>> olso.ready(model, parallel_context)
+        >>> logits = model(x)
+        >>> loss = criterion(logits, labels)
+        >>> model.backward(loss)
+    Args:
+        module (nn.Module): PyTorch module object
+        parallel_context (ParallelContext): process group object
+    """
+
+    def __init__(
+        self,
+        module: torch.nn.Module,
+        parallel_context: ParallelContext = None,
+        bucket_cap_mb: int = 25,
+        rebuild_bucket: bool = True,
+    ) -> None:
+        super().__init__(parallelism_priority=99)
+        self.module = module
+        self.parallel_context = get_parallel_context(module, parallel_context)
+        self.bucket_cap_mb = bucket_cap_mb
+        self.rebuild_bucket = rebuild_bucket
+
+    def forward(self, *args, **kwargs):
+        return self.module_forward(*args, **kwargs)
+
+    def deparallelize(self):
+        self.module.deparallelize()
+
+    def parallelize(self):
+        self.module = _DistributedDataParallel(
+            self.module, self.parallel_context, self.bucket_cap_mb, self.rebuild_bucket
+        )
+        self.module_forward = copy.copy(self.module.forward)
+
+
+class _DistributedDataParallel(nn.Module):
+    """Distributed data parallel for Oslo.
+
+    Args:
+    module (nn.Module): PyTorch module object
+    parallel_context (ParallelContext): process group object
+    """
+
+    def __init__(
+        self,
+        module: torch.nn.Module,
+        parallel_context: ParallelContext = None,
+        bucket_cap_mb: int = 25,
+        rebuild_bucket: bool = True,
+    ) -> None:
+        super().__init__()
+        self.module = module
+        self.module.zero_grad = self.zero_grad
+        self.module_forward = module.forward
+
+        self.comm_stream: torch.cuda.Stream = torch.cuda.Stream()
+        assert parallel_context
+        self.parallel_context = get_parallel_context(module, parallel_context)
+        self.dp_world_size = self.parallel_context.get_world_size(ParallelMode.DATA)
+
+        self.reducer = Reducer(bucket_cap_mb)
+        self.rebuild_bucket = rebuild_bucket
+        for p in module.parameters():
+            if is_ddp_ignored(p):
+                continue
+            if p.requires_grad:
+                p.register_hook(partial(self.grad_handle, p))
+
+    def parameters(self, recurse: bool = True):
+        return self.module.parameters(recurse)
+
+    def forward(self, *args, **kwargs):
+        self.module.zero_grad(set_to_none=True)
```

Comment:
```suggestion
```

Reply:
This was from original code.

Reply:
No, I used `model.zero_grad` out of `_DistributedDataParallel` class. It seems more natural to me.

Reply:
removed from forward